### PR TITLE
Memory-map files when possible

### LINKED ISSUE #52: TODO: Deserialize from disk by using mmap
Requires cleanup of serialization/deserialization logic so in-memory representations match the on-disk representations.  Ideally this can also be done efficiently for Fixcache entries.

### LINKED ISSUE #92: Move linking into the offline trusted toolchain
If we had a Fix jump table mapped to a known set of virtual addresses, and Runnables are otherwise position-independent, we could pre-link all the Runnables and then directly mmap them read-only (treating them the same as any other Blob). Which would let us get rid of the in-memory cache of "linked" programs.

### PR DESCRIPTION
**Pull request by Akshay-Srivatsan**
Closes #52 (more or less), but does *not* do the same to #92.  All files are deserialized and serialized via mmap.  ELF files are copied if they need to be linked, as linking involves resolving relocations (and therefore mutating data).  The biggest change here is that `RuntimeStorage` now owns all the data for any Blob or Tree, which is—to some degree—compile-time checked.  It should no longer be possible to accidentally free the data a blob was referring to without doing something obviously bad.  More incidental changes are that Tags now refer to Trees (they don't have a distinct storage type), and various uses of `span_view` have been replaced with `std::span`.

TODO: we can probably refactor some things to make "MutObject" unnecessary within RuntimeStorage; i.e., everything RuntimeStorage knows about would be immutable.

Depends on fix-project/bootstrap#3.

### CODE REVIEW

File: src/component/object.hh [hunk 1]
```diff
@@ -1,101 +1,127 @@
...
 public:
-  using ptr_type = std::unique_ptr<T, OptionalFree<T>>;
-  using view = View;
+  using span_type = S;
+  using element_type = S::element_type;
+  using value_type = S::value_type;
+  using pointer = element_type*;
+  using reference = element_type&;
+  using const_pointer = const element_type*;
+  using const_reference = element_type&;
```
Comment (yhdengh) [inline src/component/object.hh:71]: Should this be `const element_type&`?

File: src/component/object.cc [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+#include <fcntl.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+
+#include <glog/logging.h>
+
+template<any_span T>
+size_t byte_size( T& span )
+{
+  return span.size() * sizeof( typename T::element_type );
```
Comment (yhdengh) [inline src/component/object.cc:12]: Why don't we use `std::span::size_bytes`?

File: src/runtime/runtime.hh [hunk 3]
```diff
@@ -7,6 +7,7 @@
 #include "scheduler.hh"
 #include "worker_pool.hh"
 
+#define RUNTIME_THREADS 1
```
Comment (yhdengh) [inline src/runtime/runtime.hh:10]: I guess this line should be deleted?

File: src/component/handle.hh [hunk 4]
```diff
@@ -262,6 +269,26 @@ public:
     __builtin_unreachable();
   }
 
+  Handle as_tree() const { return Handle::make_tree( *this ); }
+  Handle as_thunk() const { return Handle::make_thunk( *this ); }
+  Handle as_tag() const { return Handle::make_tag( *this ); }
```
Comment (yhdengh) [inline src/component/handle.hh:274]: I wonder whether `as_tag` should be a private method...

File: src/runtime/message.hh [hunk 5]
```diff
@@ -21,65 +21,42 @@ public:
...
-  Message& operator=( Message&& ) = default;
-  Message( const Message& ) = delete;
-  Message& operator=( const Message& ) = delete;
+  /*   Message( Message&& ) = default; */
+  /*   Message& operator=( Message&& ) = default; */
+  /*   Message( const Message& ) = delete; */
+  /*   Message& operator=( const Message& ) = delete; */
 
-  virtual ~Message() {}
+  /*   virtual ~Message() {} */
```
Comment (yhdengh) [inline src/runtime/message.hh:59]: These commented out lines can just be deleted

File: src/component/handle.hh (hunk 4 above)
Comment (Akshay-Srivatsan) [reply src/component/handle.hh:281]: I'm against treating the C++ type system as a security policy given how easy to circumvent it is (and how trivial it is to write to the `.fix` directory), but I guess it's a minor enough change.  Really we should have "signed tags" like git, and the actual security guarantees are offloaded to trusting the signing key.

**Review by yhdengh — APPROVED** [review]
lgtm. But we probably should merge in #126 first?

**Pull request merged**